In [20]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [21]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id",
    encoding="utf-8")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [22]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [23]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

In [24]:
df_market_data.columns

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

### Prepare the Data

In [25]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
df_market_data_scaled = StandardScaler().fit_transform(
    df_market_data[['price_change_percentage_24h', 
                    'price_change_percentage_7d',
                    'price_change_percentage_14d',
                    'price_change_percentage_30d',
                    'price_change_percentage_60d',
                    'price_change_percentage_200d',
                    'price_change_percentage_1y']]
)

df_market_data_scaled[:1]

array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688]])

In [26]:
# Create a DataFrame with the scaled data
df_market_data_scaled_df = pd.DataFrame(df_market_data_scaled, columns=[
                    'price_change_percentage_24h', 
                    'price_change_percentage_7d',
                    'price_change_percentage_14d',
                    'price_change_percentage_30d',
                    'price_change_percentage_60d',
                    'price_change_percentage_200d',
                    'price_change_percentage_1y'
])

# Copy the crypto names from the original data
# Set the coinid column as index
df_market_data_scaled_df.index = df_market_data.index

# Display sample data
df_market_data_scaled_df.sample(5)


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                     
binance-usd                          0.061339                   -0.706669   
bitcoin-cash                         1.193036                    2.000959   
ripple                              -0.040764                   -0.810928   
ethereum-classic                    -0.071312                   -0.229484   
tron                                 0.130050                   -0.041018   

                  price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                      
binance-usd                         -0.015321                    -0.058694   
bitcoin-cash                         1.760610                     0.545842   
ripple                               0.249458                    -0.050388   
ethereum-classic                    -0.175544                     0.051882   
tron                                 0.147155                    -0.543776   

                  price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                       
binance-usd                          0.004017                     -0.550003   
bitcoin-cash                        -0.291203                     -0.499848   
ripple                              -0.373164                     -0.458259   
ethereum-classic                    -0.551760                     -0.526060   
tron                                 0.120116                     -0.241118   

                  price_change_percentage_1y  
coin_id                                       
binance-usd                        -0.281963  
bitcoin-cash                       -0.270317  
ripple                             -0.295546  
ethereum-classic                   -0.273062  
tron                               -0.234014

---

### Find the Best Value for k Using the Original Data.

In [27]:
# Create a list with the number of k-values from 1 to 11 
# was a little unclear if this wanted 10 or 11 total values. Went with the code below for a total of 10 k-values. 
k = list(range(1, 11))

In [28]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_data_scaled)
    inertia.append(k_model.inertia_)


c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environme

In [29]:
# Create a dictionary with the data to plot the Elbow curve
# Create a DataFrame with the data to plot the Elbow curve

elbow_data = {"k": k, "inertia": inertia}
df_elbow = pd.DataFrame(elbow_data)
df_elbow.head()

k     inertia
0  1  287.000000
1  2  195.820218
2  3  123.190482
3  4   79.022435
4  5   65.302379

In [30]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.

df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value for 'k' appears to lie somewhere between 3 and 4. The first apparent elbow is present at 3, though another is visable at 4. I would begin my analysis with a clustering of 4.

When completing the graph below with a cluster of 4, you see a couple of potential outliers, and two distinct clusters. When using 3, you have two outliers and one main cluster. 

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [31]:
# Initialize the K-Means model using the best value for k# Define the model with optimal number of clusters
model = KMeans(n_clusters=4, random_state=0)

In [32]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_scaled)


c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [33]:
# Predict the clusters to group the cryptocurrencies using the scaled data
kmeans_predictions = model.predict(df_market_data_scaled)

# Print the resulting array of cluster values.
kmeans_predictions



array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 1, 0, 3, 3, 2, 3, 3, 3, 3])

In [34]:
# Create a copy of the DataFrame
df_market_data_predictions = df_market_data_scaled_df.copy()

df_market_data_predictions.sample(5)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                     
binancecoin                          0.891871                    1.327295   
bitcoin-cash-sv                      0.448952                   -0.190684   
crypto-com-chain                     0.331280                   -1.614844   
huobi-token                         -0.052032                   -0.457229   
neo                                  0.286546                   -0.326301   

                  price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                      
binancecoin                          0.800214                    -0.057148   
bitcoin-cash-sv                     -0.248043                     0.051634   
crypto-com-chain                    -1.054521                    -0.729931   
huobi-token                          0.032522                    -0.184489   
neo                                 -1.212670                    -0.903134   

                  price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                       
binancecoin                          0.778653                     -0.188232   
bitcoin-cash-sv                     -0.529666                     -0.532961   
crypto-com-chain                    -0.350155                     -0.022866   
huobi-token                         -0.070809                     -0.451982   
neo                                  0.290970                     -0.181187   

                  price_change_percentage_1y  
coin_id                                       
binancecoin                        -0.225533  
bitcoin-cash-sv                    -0.206029  
crypto-com-chain                   -0.034570  
huobi-token                        -0.262140  
neo                                -0.175550

In [35]:
# Add a new column to the DataFrame with the predicted clusters
df_market_data_predictions["clusters"] = kmeans_predictions

# Display sample data
df_market_data_predictions.sample(5)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
ftx-token                           0.414711                    0.414044   
wrapped-bitcoin                     0.515453                    0.461843   
cosmos                             -0.255978                    1.840274   
ontology                           -0.409044                   -0.906963   
cardano                             0.102530                    1.508001   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
ftx-token                          -0.047386                    -0.465380   
wrapped-bitcoin                     0.769975                     0.224045   
cosmos                              0.643565                     0.116538   
ontology                           -1.298986                    -1.393153   
cardano                             0.648885                     0.328959   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
ftx-token                           0.128185                     -0.416502   
wrapped-bitcoin                    -0.074674                     -0.355054   
cosmos                             -0.151913                     -0.117565   
ontology                           -0.696937                     -0.422835   
cardano                            -0.486349                      0.065080   

                 price_change_percentage_1y  clusters  
coin_id                                                
ftx-token                         -0.145469         3  
wrapped-bitcoin                   -0.251623         0  
cosmos                            -0.215191         0  
ontology                          -0.292344         3  
cardano                           -0.155428         0

In [36]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

df_market_data_predictions.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    hover_cols='coin_id',
    by="clusters"
)

:NdOverlay   [clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [38]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)


In [45]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
df_market_data_scaled_pca = pca.fit_transform(df_market_data_scaled_df)

# View the first five rows of the DataFrame. 
df_market_data_scaled_pca[:5]


array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [51]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
exp_variance = pca.explained_variance_ratio_

sum(exp_variance)


0.895031657030984

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance of the three principal components is 0.895031657030984

In [56]:
# Create a new DataFrame with the PCA data.
market_data_pca_df = pd.DataFrame(df_market_data_scaled_pca, columns=["PCA1", "PCA2", "PCA3"])

# Copy the crypto names from the original data
# Set the coinid column as index
market_data_pca_df.index = df_market_data.index

# Display sample data
market_data_pca_df.sample(10)


PCA1      PCA2      PCA3
coin_id                                   
chainlink    -0.450711  0.517699  2.846143
neo          -0.362120 -0.986914 -0.728752
digibyte     -0.297910 -0.191126 -0.909602
bitcoin-cash -1.157800  2.041209  1.859715
ripple       -0.471835 -0.222660 -0.479053
litecoin     -0.649468  0.432165  0.600303
ethereum     -0.458261  0.458466  0.952877
nem           0.608974  0.563532 -1.148742
tezos        -0.796176 -0.494409  1.082812
usd-coin     -0.438408 -0.175337 -0.663388

---

### Find the Best Value for k Using the PCA Data

In [61]:
# Create a list with the number of k-values from 1 to 11
pca_k = list(range(1, 11))

In [62]:
# Create an empty list to store the inertia values
pca_inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    pca_k_model = KMeans(n_clusters=i, random_state=0)
    pca_k_model.fit(market_data_pca_df)
    pca_inertia.append(pca_k_model.inertia_)

c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environme

In [63]:
# Create a dictionary with the data to plot the Elbow curve
# Create a DataFrame with the data to plot the Elbow curve
pca_elbow_data = {"k": pca_k, "inertia": pca_inertia}
pca_df_elbow = pd.DataFrame(pca_elbow_data)

pca_df_elbow

k     inertia
0   1  256.874086
1   2  165.901994
2   3   93.774626
3   4   49.665497
4   5   37.878747
5   6   27.618972
6   7   21.182776
7   8   17.389823
8   9   13.593650
9  10   10.559358

In [64]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
pca_df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="PCA Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** It appears as though the best value for K when using the PCA data is 4. The 


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** 

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [65]:
# Initialize the K-Means model using the best value for k
pca_model = KMeans(n_clusters=4, random_state=0)

In [66]:
# Fit the K-Means model using the PCA data
pca_model.fit(market_data_pca_df)

c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [68]:
# Predict the clusters to group the cryptocurrencies using the PCA data

pca_predictions = pca_model.predict(market_data_pca_df)
# Print the resulting array of cluster values.
pca_predictions

array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 1, 0, 3, 3, 2, 3, 3, 3, 3])

In [71]:
# Create a copy of the DataFrame with the PCA data
market_data_pca_pred = market_data_pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
market_data_pca_pred['clusters'] = pca_predictions

# Display sample data
market_data_pca_pred.sample(5)


PCA1      PCA2      PCA3  clusters
coin_id                                            
okb          0.064075 -1.269825 -1.098829         3
binance-usd -0.450211 -0.151019 -0.647401         3
cdai        -0.513528 -0.142802 -0.656566         3
chainlink   -0.450711  0.517699  2.846143         0
ripple      -0.471835 -0.222660 -0.479053         3

In [75]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

market_data_pca_pred.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    hover_cols='coin_id',
    by="clusters"
)

:NdOverlay   [clusters]
   :Scatter   [PCA2]   (PCA3,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [30]:
# Composite plot to contrast the Elbow curves

In [31]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 